### ISSUE TITLE
[Bug]: Rich Text Editor Issues

### ISSUE THREAD

**Issue by riodeuno** [appsmithorg/appsmith#12321]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

There are a few nitpicks and inconsistencies in the Rich Text Editor.

1. `RichTextEditor1.text` binding value has a strange experience. This could be due to how the Text Widget renders the HTML provided by the `RichTextEditor1.text`. Observations from the video:
 - The HTML formatting stops working after a few modifications
 - Linebreaks don't show up in HTML (both in the Text Widget as well as the RTE widget, see video)

https://user-images.githubusercontent.com/103687/160553060-6f647adc-8f50-4241-8fdd-d9bd8ba650ca.mov

Uploading Screen Recording 2022-03-29 at 12.14.02 PM.mov…

2. Also, the `InputType` label on the RTE property pane seems misleading. We can alternatively call it "Default Text Type/Markup Type"

### Steps To Reproduce

1. Add a RTE widget.
2. Select InputType as "Markdown"
3. Add some markdown content
4. Bind the `RichTextEditor1.text` property to a Text widget
5. Try adding linebreaks into the markdown default content.

### Version

Cloud / All

**Comment by dilippitchika** [comment]
@riodeuno thanks for pointing. 

@vivekverma2312 i think this needs deeper investigation to identify all such inconsistencies across the RTE, can we please assign someone to dig a bit deeper around these problems?

**Comment by sbres** [comment]
I don't know if it helps but when getting the text from a rich text editor in Markdown mode
`RichTextEditor1.text`
The resulting text is in an HTML form that destroys the purpose of using the editor as destroys all the markdown code my input

**Comment by riodeuno** [comment]
@sbres Thanks for pointing this scenario!
By design, today, the Rich Text Editor, exposes its contents in HTML. The Markdown and HTML formats for default input, is actually specifically for the format in which we would like to provide default values to the rich text editor.

Having said that, I understand the need to be able to edit Rich Text and export the result in Markdown. The status of this feature, can be tracked here: #2047

**Comment by shadabbuchh** [comment]
1. Missing "Center" Alignment for Label: At present we have left and right alignments for the label in RTE widget. Should we also introduce a Center alignment?
![Alignment_01.PNG](https://images.zenhubusercontent.com/6278b12bad578328630ba2e7/449941c1-c275-4c46-8864-6a49c8ab39d6)
2. Line breaks do not show up in the Label
![Line Break 02.PNG](https://images.zenhubusercontent.com/6278b12bad578328630ba2e7/5e1f1ff4-a2f0-484f-8efb-1e76780fa34b)
3. Refresh issue: When the input type is changed form `Markdown` to `HTML` or vice versa, the format of the text in the editor does not change unless a hard refresh is done.
https://images.zenhubusercontent.com/275993885/5fd47918-c7ba-4026-bab5-ad3afcc5f59c/my_first_application___editor___appsmith___mozilla_firefox_2022_05_11_15_19_05_trim.mp4

**Comment by sbres** [comment]
Thanks for your answer @riodeuno It seems the issue has been there for one year and is not moving forward "/

**Comment by dilippitchika** [comment]
Hey @sbres apologies, I have a question 
-  Are you looking for the output from `richTextEditor.text` to be in a markdown format? using the markdown tags than the html tags?

**Comment by dilippitchika** [comment]
Hey @shadabbuchh can you check this, I discussed with Tolu and this is also a bug - https://www.loom.com/share/833ebfc5e1f14cc0bff983e28d19d796

**Comment by shadabbuchh** [comment]
> Hey @shadabbuchh can you check this, I discussed with Tolu and this is also a bug - [![LOOM DEMO](https://camo.githubusercontent.com/bd1d6bac5ef0cfe974b8690e7e72fb91d5ed618afacd49c80b4695b97eb734c6/687474703a2f2f63646e2e6c6f6f6d2e636f6d2f73657373696f6e732f7468756d626e61696c732f38333365626663356531663134636330626666393833653238643139643739362d30303030312e676966)](https://www.loom.com/share/833ebfc5e1f14cc0bff983e28d19d796)

Yes this seems to be an issue as well. Thanks you for pointing it out.

**Comment by shadabbuchh** [comment]
@dilippitchika yes this seems to be an issue as well. Thank you for pointing it out.

**Comment by sbres** [comment]
@dilippitchika Yes in fact that's what I expected.
I'm imputing markdown text into it, I was expecting to have markdown out also.
The fact that it outputs HTML, it makes it totally unuseful as I can't store the output anymore.
I'm not using this element anymore until I can get markdown out.

**Comment by dilippitchika** [comment]
Understood, this looks more like a bug. Let me try to pick this up on priority

**Comment by dilippitchika** [comment]
Hey @sbres turns out the library we are using only supports html output. We are tracking this in a new issue and will pick this up sooner - https://github.com/appsmithorg/appsmith/issues/13785

In the meantime if you are familiar with JS objects you can try to implement a js function which does this for you. Here's a JS snippet if you are interested. I will try to create a sample app from this and share. - https://jonlabelle.com/snippets/view/javascript/convert-html-into-markdown-with-javascript

**Comment by shadabbuchh** [comment]
Some more issues:

1. On clicking a link in the RTE, the user is not redirected to it either on the current window or in a new window. Even while previewing the same thing happens. 
https://www.loom.com/share/9c06cf1915e34b58a77767682233205e
2. Why do we have 2 text colour background colour pickers? 
![Screenshot 2022-05-23 at 5.11.20 PM.png](https://images.zenhubusercontent.com/6278b12bad578328630ba2e7/e7996cbb-7fe6-429f-a53a-4bd9a62a50da)
3. Indentation does not work on a line basis (for a line where the cursor is placed). The whole text moves irrespective of the cursor placement.
https://www.loom.com/share/00961944577444d292abdf3e2c4708c0

**Comment by dilippitchika** [comment]
**BUG** We have 2 text colour background colour pickers 

**BUG** We have inconsistent behaviour when formatting in tiny mce and formatting in our widget
Loom videos for reference - 
1. `https://www.loom.com/share/00961944577444d292abdf3e2c4708c0`
2. `https://www.loom.com/share/833ebfc5e1f14cc0bff983e28d19d796`


Not bugs
1. On clicking a link in the RTE, the user is not redirected to it either on the current window or in a new window. Even while previewing the same thing happens. 
     - Not a bug IMO, RTE is not a text widget so user will not automatically redirect to the link on click. They can see the access on right click or ctrl+click. Standard RTE behavior. If you want to test if linking is working fine, you have to check it using {{RTE.text}} in a text widget.
2. Missing "Center" Alignment for Label: At present we have left and right alignments for the label in RTE widget. Should we also introduce a Center alignment?
     - Not a bug, the right and left alignment help you align the RTE in a form when there are multiple widgets. Haven't seen the labels to be center aligned yet.
3. Line breaks don't show up in label
     - Not a bug, maybe a separate enhancement request. We never supported multi-line requests in labels. Will hold off creating a request until we get a real use case.

**Comment by dilippitchika** [comment]
@shadabbuchh can you replicate the first issue mentioned comment mentioned by Abhinav? I was unable to reproduce it clearly.

**Comment by shadabbuchh** [comment]
@dilippitchika this is what I could replicate:

- Linebreaks don't show up both in the Text Widget as well as the RTE widget.
- After introducing line breaks, the hyperlinks change.

https://www.loom.com/share/0ce0db441dc74cc2992127fc55663ca5

Not sure what Abhinav meant by `The HTML formatting stops working after a few modifications`.

**Comment by dilippitchika** [comment]
> Not sure what Abhinav meant by The HTML formatting stops working after a few modifications.

Hey Shadab, if you keep an eye on the text widget in your video you can see the formatting change after you add line breaks. You start seeing ### instead of the heading be large etc. That is what abhinav meant

**Comment by shadabbuchh** [comment]
> Can you also add the text you pasted as markdown in the RTE widget here?

```
---
__Advertisement :)__

- __[pica](https://nodeca.github.io/pica/demo/)__ - high quality and fast image
  resize in browser.
- __[babelfish](https://github.com/nodeca/babelfish/)__ - developer friendly
  i18n with plurals support and easy syntax.

You will like those projects!

---

# h1 Heading 8-)
## h2 Heading
### h3 Heading
#### h4 Heading
##### h5 Heading
###### h6 Heading


## Horizontal Rules
```

**Comment by shadabbuchh** [comment]
Some more issues with the RTE widget:

1. Bullet points don't start from where the cursor is pointing.
2. The Tab/Indent icon shifts the upper line/paragraph as well.
3. Link automatically gets highlighted. Can't enter any text after the link in the same line.

https://www.loom.com/share/20014ef063f743e396357122edcb3b00

**Comment by Shivaansh-Agarwal** [comment]
Facing another issue with Rich Text Editor:

1. When making the RTE disabled, it doesn't allow the content to be scrolled, even though I've made it auto-height.

<img width="1415" alt="image" src="https://user-images.githubusercontent.com/29867531/209645911-f4a27b54-5e4d-404d-a2a1-ec0157a12b24.png">
